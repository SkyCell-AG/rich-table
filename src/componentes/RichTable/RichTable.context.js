import {
    createContext,
} from 'react'
import noop from 'lodash/noop'

const RichTableContext = createContext({
    setDataTable: noop,
})

export default RichTableContext
