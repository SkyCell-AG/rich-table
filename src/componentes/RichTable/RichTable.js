import React, {
    useCallback,
    forwardRef,
    useEffect,
    useState,
    useRef,
} from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import noop from 'lodash/noop'
import get from 'lodash/get'

import HeaderCell from 'componentes/HeaderCell'
import InfiniteList from 'componentes/InfiniteList'
import DndCell from 'componentes/DndCell'
import BaseRow from 'componentes/BaseRow'

import useStyles from './RichTable.style'

const propTypes = {
    uniqField: PropTypes.string,
    name: PropTypes.string.isRequired,
    configName: PropTypes.string,
    load: PropTypes.func.isRequired,
    visible: PropTypes.arrayOf(PropTypes.string).isRequired,
    setVisible: PropTypes.func.isRequired,
    onRowClick: PropTypes.func,
    changeSequence: PropTypes.func.isRequired,
    setMatchedResults: PropTypes.func.isRequired,
    matchedResults: PropTypes.number,
    removeFilter: PropTypes.func.isRequired,
    removeSort: PropTypes.func.isRequired,
    rerenderInfinitList: PropTypes.func.isRequired,
    infinitListKey: PropTypes.string.isRequired,
    selectedRows: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    columns: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            Header: PropTypes.func,
            Cell: PropTypes.func,
            props: PropTypes.object, // eslint-disable-line
        }).isRequired,
    ).isRequired,
    className: PropTypes.string,
    controlPanel: PropTypes.func,
    renderControlPanel: PropTypes.func,
    classNames: PropTypes.exact({
        root: PropTypes.string,
        contentWrapper: PropTypes.string,
        headerWrapper: PropTypes.string,
    }),
    editing: PropTypes.bool,
    selectedRowId: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]),
    detailPanel: PropTypes.func,
    openRowId: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]),
    rowStyle: PropTypes.shape({
        crossLine: PropTypes.string,
    }),
    disabled: PropTypes.bool,
}

const defaultProps = {
    matchedResults: undefined,
    className: '',
    onRowClick: undefined,
    selectedRows: undefined,
    uniqField: 'id',
    controlPanel: null,
    configName: undefined,
    renderControlPanel: null,
    classNames: {},
    editing: false,
    selectedRowId: '',
    detailPanel: undefined,
    openRowId: undefined,
    rowStyle: undefined,
    disabled: false,
}

const RichTable = forwardRef((props, ref) => {
    const {
        columns,
        className: wrapperClassName,
        setMatchedResults,
        onRowClick,
        selectedRows,
        infinitListKey,
        changeSequence,
        uniqField,
        controlPanel: ControlPanel,
        renderControlPanel,
        classNames,
        editing,
        selectedRowId,
        name,
        configName,
        detailPanel,
        rerenderInfinitList,
        openRowId,
        load,
        rowStyle,
        disabled,
    } = props

    const classes = useStyles()

    const [
        openedRow,
        setOpenedRow,
    ] = useState(null)

    const closeOpenedRow = useCallback(() => {
        setOpenedRow(null)
    }, [])

    const rowClick = useCallback((row) => {
        return () => {
            if (disabled) { return }
            if (onRowClick) {
                onRowClick(
                    row,
                    rerenderInfinitList,
                )
            } else if (detailPanel) {
                setOpenedRow(openedRow === row ? null : row)
            }
        }
    }, [
        openedRow,
        onRowClick,
        detailPanel,
        rerenderInfinitList,
        disabled,
    ])

    const selectedElm = useRef()

    useEffect(() => {
        if (selectedRowId && selectedElm.current) {
            if (selectedElm.current.scrollIntoViewIfNeeded) {
                selectedElm.current.scrollIntoViewIfNeeded()
            } else {
                // IE, Firefox
                selectedElm.current.scrollIntoView()
            }
        }
    }, [selectedRowId])

    const renderedControlPanel = (renderControlPanel) ? renderControlPanel(props) : null

    return (
        <div
            className={clsx(
                classNames.root,
                classes.tableWrapper,
                editing && classes.tableWrapperWhenEditing,
                wrapperClassName,
            )}
        >
            <div
                className={clsx(
                    classes.headerTableWrapper,
                    editing && classes.tableWrapperWhenEditing,
                    classNames.headerWrapper,
                )}
            >
                {ControlPanel && (
                    <ControlPanel
                        {...props}
                    />
                )}
                {renderedControlPanel}
            </div>

            <div
                className={clsx(
                    classes.contentWrapper,
                    editing && classes.contentWrapperWhenEditing,
                    classNames.contentWrapper,
                )}
            >
                <InfiniteList
                    {...props}
                    load={load}
                    ref={ref}
                    key={infinitListKey}
                    onUpdateMatchedResults={setMatchedResults}
                    BeforeList={({
                        data,
                    }) => {
                        return (
                            <div className={clsx(
                                classes.headings,
                                {
                                    [classes.withDetailPanel]: detailPanel,
                                },
                            )}
                            >
                                {columns.map(({
                                    Header = HeaderCell,
                                    mapHeaderProps = noop,
                                    headerProps: headerPropsOut,
                                    filterField,
                                    id,
                                    className: cellClassName,
                                    width,
                                    minWidth = '',
                                    props: columnProps,
                                    ...rest
                                }) => {
                                    const headerProps = headerPropsOut
                                        || mapHeaderProps({
                                            ...columnProps,
                                            data,
                                        })

                                    const divStyles = width ? {
                                        flex: `0 0 ${width}`,
                                        minWidth: `${minWidth}`,
                                    } : {
                                        minWidth: `${minWidth}`,
                                    }

                                    return (
                                        <div
                                            key={`DndCell-${id}`}
                                            data-testid={`table-header-${id}`}
                                            className={clsx(
                                                classes.cell,
                                                cellClassName,
                                            )}
                                            style={divStyles}
                                        >
                                            <DndCell
                                                {...headerProps}
                                                onDrop={changeSequence}
                                                id={id}
                                                type="HeaderCell"
                                                data={data}
                                            >
                                                <Header
                                                    {...rest}
                                                    {...columnProps}
                                                    {...headerProps}
                                                    data={data}
                                                    key={`header-${id}`}
                                                    id={filterField || id}
                                                />
                                            </DndCell>
                                        </div>
                                    )
                                })}
                            </div>
                        )
                    }}
                    Row={(rowProps) => {
                        const uniqFieldValue = get(rowProps, uniqField)

                        const innerOpenRowId = get(openedRow, uniqField)

                        const openRow = (
                            openRowId === uniqFieldValue
                            || innerOpenRowId === uniqFieldValue
                        )

                        return (
                            <div
                                ref={selectedRowId === uniqFieldValue ? selectedElm : null}
                                key={`row-${uniqFieldValue}`}
                            >
                                <BaseRow
                                    name={name}
                                    configName={configName}
                                    selectedRows={selectedRows}
                                    selectedRowId={selectedRowId}
                                    uniqFieldValue={uniqFieldValue}
                                    rowClick={rowClick(rowProps)}
                                    rowProps={rowProps}
                                    columns={columns}
                                    detailPanel={detailPanel}
                                    openRow={openRow}
                                    closeOpenedRow={closeOpenedRow}
                                    rowStyle={rowStyle}
                                    disabled={disabled}
                                />
                            </div>
                        )
                    }}
                />
            </div>
        </div>
    )
})

RichTable.propTypes = propTypes
RichTable.defaultProps = defaultProps

export default RichTable
