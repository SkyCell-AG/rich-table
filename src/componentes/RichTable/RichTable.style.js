import {
    makeStyles,
} from '@material-ui/styles'

const useStyles = makeStyles({
    headings: {
        position: 'sticky',
        zIndex: 1,
        top: 0,
        display: 'flex',
        justifyContent: 'space-around',
        boxShadow: (theme) => { return `0 1px 2px 0 ${theme.palette.secondary[100]}` },
        minHeight: (theme) => { return theme.shape.tabHeight },
        backgroundColor: (theme) => { return theme.palette.common.white },
    },
    tableWrapper: {
        position: 'relative',
        width: '100%',
        display: 'flex',
        flexDirection: 'column',
        fontSize: '1rem',
    },
    tableWrapperWhenEditing: {
        marginRight: 0,
    },
    tabs: {
        display: 'flex',
        alignItems: 'center',
    },
    headerTableWrapper: {
        display: 'flex',
        position: 'absolute',
        width: '100%',
        overflow: 'hidden',
        overflowX: 'auto',
        justifyContent: 'space-between',
    },
    controlsWrapper: {
        display: 'flex',
        justifyContent: 'right',
        maxHeight: (theme) => { return theme.shape.tabHeight },
    },
    contentWrapperWhenEditing: {
        borderRight: 'none !important',
    },
    controls: {
        display: 'flex',
        zIndex: 1,
        height: (theme) => { return theme.shape.tabHeight },
        backgroundColor: (theme) => { return theme.palette.secondary[50] },
        paddingLeft: 10,
        whiteSpace: 'nowrap',
        borderRadius: '10px 10px 0px 0px',
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: (theme) => { return theme.palette.secondary[300] },
        borderBottom: (theme) => { return `1px solid ${theme.palette.secondary[50]}` },
    },
    control: {
        cursor: 'pointer',
        display: 'flex',
        alignItems: 'center',
        color: (theme) => { return theme.palette.secondary[600] },
        padding: '2px 20px 2px 0px',
    },
    contentWrapper: {
        position: 'relative',
        left: 0,
        padding: (theme) => { return theme.spacing(1.5) },
        top: (theme) => { return theme.shape.tabHeight - 1 },
        width: (theme) => { return `calc(100% - ${theme.spacing(3)}px)` },
        height: (theme) => { return `calc(100% - ${theme.shape.tabHeight + theme.spacing(3) + 2}px)` },
        borderRadius: '10px 0 0 0',
        backgroundColor: (theme) => { return theme.palette.secondary[50] },
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: (theme) => { return theme.palette.secondary[300] },
    },
    cell: {
        width: '100%',
        overflow: 'hidden',
    },
    cursorPointer: {
        cursor: 'pointer',
    },
    withDetailPanel: {
        paddingLeft: 24,
    },
})

export default useStyles
