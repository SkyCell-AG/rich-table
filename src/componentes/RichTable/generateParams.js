import memoize from 'lodash/memoize'
import isEmpty from 'lodash/isEmpty'

const generateParams = memoize((filter, sort, excludeFilters) => {
    const params = {}

    if (!isEmpty(filter)) {
        params.filter = filter
    }

    if (!isEmpty(sort)) {
        params.sort = sort
    }
    if (!isEmpty(excludeFilters)) {
        params.excludeFilters = excludeFilters
    }

    return params
}, (...data) => {
    return JSON.stringify(data)
})

export default generateParams
