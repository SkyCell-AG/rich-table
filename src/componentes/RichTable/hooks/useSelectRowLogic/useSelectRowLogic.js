import {
    useMemo,
} from 'react'

import useSelectRow from '../useSelectRow'
import useRadioSelectRow from '../useRadioSelectRow'

const useSelectRowLogic = ({
    columns,
    onSelectRow,
    uniqField,
    selectedRows = [],
    radioSelect,
    disabled,
    dataTable,
    showSelectAll,
}) => {
    const visibleAndSortedColumnsWithRadioSelection = useRadioSelectRow({
        columns,
        onSelectRow,
        uniqField,
        selectedRows,
        disabled,
    })

    const visibleAndSortedColumnsWithSelection = useSelectRow({
        columns,
        onSelectRow,
        uniqField,
        selectedRows,
        disabled,
        dataTable,
        showSelectAll,
    })

    return useMemo(() => {
        if (!onSelectRow) {
            return columns
        }

        if (radioSelect) {
            return visibleAndSortedColumnsWithRadioSelection
        }

        return visibleAndSortedColumnsWithSelection
    }, [
        columns,
        onSelectRow,
        radioSelect,
        visibleAndSortedColumnsWithRadioSelection,
        visibleAndSortedColumnsWithSelection,
    ])
}

export default useSelectRowLogic
