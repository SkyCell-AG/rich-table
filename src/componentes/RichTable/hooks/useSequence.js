import {
    useEffect, useReducer, useCallback,
} from 'react'

import createReducer from 'utils/createReducer'

const SET_COLUMNS_SEQUENCE = 'SET_COLUMNS_SEQUENCE'
const CHANGE_COLUMNS_SEQUENCE = 'CHANGE_COLUMNS_SEQUENCE'

const reducer = createReducer({
    [SET_COLUMNS_SEQUENCE]: (_, {
        meta: {
            sequence,
        },
    }) => {
        return sequence
    },
    [CHANGE_COLUMNS_SEQUENCE]: (sequence, {
        meta: [
            a,
            b,
        ],
    }) => {
        const aIndex = sequence.indexOf(a)
        const sequenceWithoutB = sequence.filter((elem) => {
            return elem !== b
        })

        return [
            ...sequenceWithoutB.slice(0, aIndex),
            b,
            ...sequenceWithoutB.slice(aIndex),
        ]
    },
})

const useSequence = (columns) => {
    const [
        sequence,
        dispatch,
    ] = useReducer(reducer, [])

    useEffect(() => {
        dispatch({
            type: SET_COLUMNS_SEQUENCE,
            meta: {
                sequence: columns.map(({
                    id,
                }) => {
                    return id
                }),
            },
        })
    }, [columns])

    const changeSequence = useCallback((a, b) => {
        dispatch({
            type: CHANGE_COLUMNS_SEQUENCE,
            meta: [
                a,
                b,
            ],
        })
    }, [dispatch])

    return [
        sequence,
        changeSequence,
    ]
}

export default useSequence
