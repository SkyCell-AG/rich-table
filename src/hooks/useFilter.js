import {
    useReducer,
    useCallback,
    useEffect,
} from 'react'

import createReducer from 'utils/createReducer'
import removeEmptyFields from 'utils/removeEmptyFields'

const UPDATE_FILTER = 'UPDATE_FILTER'
const SET_FILTER = 'SET_FILTER'
const REMOVE_FILTER = 'REMOVE_FILTER'

const initState = {
    filter: {},
    predefinedFilter: {},
}

const reducer = createReducer({
    [REMOVE_FILTER]: ({
        predefinedFilter,
    }) => {
        return {
            ...initState,
            filter: predefinedFilter,
            predefinedFilter,
        }
    },
    [UPDATE_FILTER]: (state, {
        meta: {
            filter,
        },
    }) => {
        return {
            ...state,
            filter,
            predefinedFilter: filter,
        }
    },
    [SET_FILTER]: (state, {
        meta: {
            column, value,
        },
    }) => {
        return {
            ...state,
            filter: removeEmptyFields({
                ...state.filter,
                [column]: value,
            }),
        }
    },
})

const useFilter = (filterOut) => {
    const [
        {
            filter,
            predefinedFilter,
        },
        dispatch,
    ] = useReducer(reducer, {
        filter: filterOut,
        predefinedFilter: filterOut,
    })

    useEffect(() => {
        if (predefinedFilter !== filterOut) {
            dispatch({
                type: UPDATE_FILTER,
                meta: {
                    filter: filterOut,
                },
            })
        }
    }, [
        predefinedFilter,
        filterOut,
    ])

    const setFilter = useCallback((column, value) => {
        dispatch({
            type: SET_FILTER,
            meta: {
                column,
                value,
            },
        })
    }, [dispatch])

    const removeFilter = useCallback(() => {
        dispatch({
            type: REMOVE_FILTER,
        })
    }, [])

    return [
        filter,
        setFilter,
        removeFilter,
    ]
}

export default useFilter
