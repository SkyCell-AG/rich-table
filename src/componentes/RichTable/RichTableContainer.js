import React, {
    useReducer,
    useCallback,
    useMemo,
    useEffect,
} from 'react'
import PropTypes from 'prop-types'
import omit from 'lodash/omit'
import {
    v1 as uuid,
} from 'uuid'

import generateFilters from 'utils/generateFilters'
import reducer from '/store/reducer'
import * as statuses from 'utils/requestStatuses'
import generateParams from 'utils/generateParams'

import useLoadPage from 'hooks/useLoadPage'
import useVisible from 'hooks/useVisible'
import useSequence from 'hooks/useSequence'
import useFilter from 'hooks/useFilter'
import useSort from 'hooks/useSort'
import useSelectRow from 'hooks/useSelectRow'

import RichTable from './RichTable'

const initialState = {
    status: statuses.PRISTIN,
    data: [],
    matchedResults: 0,
    infinitListKey: uuid(),
}

const propTypes = {
    uniqField: PropTypes.string,
    visible: PropTypes.arrayOf(PropTypes.string),
    selectedRows: PropTypes.arrayOf(PropTypes.string),
    onSelectRow: PropTypes.func,
    load: PropTypes.func.isRequired,
    name: PropTypes.string.isRequired,
    onFilterChange: PropTypes.func,
    selectedFilter: PropTypes.object, // eslint-disable-line
    filter: PropTypes.object, // eslint-disable-line
    sort: PropTypes.object, // eslint-disable-line
    columns: PropTypes.arrayOf(PropTypes.shape({
        props: PropTypes.object, // eslint-disable-line
    })).isRequired,
}

const defaultProps = {
    sort: {},
    visible: [],
    filter: {},
    uniqField: 'id',
    onSelectRow: undefined,
    onFilterChange: undefined,
    selectedRows: undefined,
}

const RichTableContainer = ({
    columns: outColumns,
    load,
    name,
    selectedRows,
    visible: predefinedVisible,
    sort: predefinedSort,
    filter: predefinedFilter,
    onSelectRow,
    uniqField,
    namedQuery,
    ...props
}) => {
    const [
        state,
        dispatch,
    ] = useReducer(reducer, initialState)

    const  {
        infinitListKey,
        matchedResults,
    } = state

    const typeMapping = useMemo(() => {
        return outColumns.reduce((prev, {
            filterField,
            filterType,
        }) => {
            if (!filterField) {
                return prev
            }

            return {
                ...prev,
                [filterField]: filterType,
            }
        }, {})
    }, [outColumns])

    const columns = useMemo(() => {
        return outColumns.map((column) => {
            const {
                mapHeaderProps,
                headerProps,
            } = column

            if (headerProps) {
                return {
                    ...column,
                    label: headerProps.children,
                }
            }

            if (mapHeaderProps) {
                return {
                    ...column,
                    label: mapHeaderProps().children,
                }
            }

            return {
                ...column,
                label: '',
            }
        })
    }, [outColumns])

    const [
        sequence,
        changeSequence,
    ] = useSequence(columns)
    const [
        visible,
        setVisible,
    ] = useVisible(columns, predefinedVisible)
    const [
        filter,
        setFilter,
        removeFilter,
    ] = useFilter(predefinedFilter)
    const [
        sort,
        setSorting,
        removeSort,
    ] = useSort(predefinedSort)

    const params = useMemo(() => {
        return generateParams(filter, sort)
    }, [
        filter,
        sort,
    ])

    const loadWithParams = useCallback((page) => {
        return load({
            ...params,
            typeMapping,
            page,
        }).catch((err) => {
            // eslint-disable-next-line no-console, max-len
            console.error(err) // TODO: without catch handler catch handler in InfinitListContainer doesn't fire
        })
    }, [
        load,
        params,
        typeMapping,
    ])

    const {
        loadPage,
        handleSetMatchedResults,
        handleRerenderInfinitList,
    } = useLoadPage(loadWithParams, dispatch)

    const createFilterHandler = useCallback((columnId) => {
        return (value) => {
            setFilter(columnId, value)
        }
    }, [setFilter])

    const visibleAndSortedColumns = sequence
        .filter((columnId) => {
            return visible.includes(columnId)
        })
        .map((columnId) => {
            return columns.find(({
                id,
            }) => {
                return id === columnId
            })
        })
        .filter(Boolean)
        .map((column) => {
            const appliedFilters = generateFilters(omit(
                filter,
                [column.filterField],
            ), typeMapping)

            return {
                ...column,
                props: {
                    appliedFilters,
                    filter: filter[column.filterField],
                    predefinedFilter,
                    setFilter: createFilterHandler(column.filterField || column.id),
                    [sort.field === column.id && 'sortDirection']: sort.direction,
                    sort: (direction) => {
                        setSorting(
                            column.id,
                            direction,
                        )
                    },
                },
            }
        })

    const visibleAndSortedColumnsWithSelection = useSelectRow({
        columns: visibleAndSortedColumns,
        onSelectRow,
        uniqField,
    })

    useEffect(() => {
        loadPage([], 1)
    }, [filter, namedQuery])

    return (
        <RichTable
            {...props}
            {...state}
            name={name}
            load={loadPage}
            filter={filter}
            removeFilter={removeFilter}
            rerenderInfinitList={handleRerenderInfinitList}
            removeSort={removeSort}
            setMatchedResults={handleSetMatchedResults}
            changeSequence={changeSequence}
            visible={visible}
            setVisible={setVisible}
            allColumns={columns}
            infinitListKey={infinitListKey}
            matchedResults={matchedResults}
            selectedRows={selectedRows}
            columns={visibleAndSortedColumnsWithSelection}
            uniqField={uniqField}
        />
    )
}

RichTableContainer.propTypes = propTypes
RichTableContainer.defaultProps = defaultProps

export default RichTableContainer
