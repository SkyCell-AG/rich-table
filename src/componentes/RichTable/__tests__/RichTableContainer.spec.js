import React from 'react'
import {
    shallow,
} from 'enzyme'

import {
    v1 as uuid,
} from 'uuid'

import useSequence from '../hooks/useSequence'
import useVisible from '../hooks/useVisible'
import useFilter from '../hooks/useFilter'
import useSort from '../hooks/useSort'
import useSelectRowLogic from '../hooks/useSelectRowLogic'
import RichTableContainer from '../RichTableContainer'
import RichTable from '../RichTable'

jest.mock('uuid', () => {
    return {
        v1: jest.fn().mockReturnValue('some-random-key'),
    }
})
jest.mock('../hooks/useSequence', () => {
    return jest.fn()
})
jest.mock('../hooks/useVisible', () => {
    return jest.fn()
})
jest.mock('../hooks/useFilter', () => {
    return jest.fn()
})
jest.mock('../hooks/useSort', () => {
    return jest.fn()
})
jest.mock('../hooks/useSelectRowLogic', () => {
    return jest.fn().mockImplementation(({
        columns: cp,
    }) => {
        return cp
    })
})

let richTableContainerElement
const columns = [
    {
        id: '1',
        filterField: 'someField',
        filterType: 'specialFilter',
        headerProps: {
            children: 'Heading 1',
        },
        props: {
            prop1: 'prop1-1',
        },
    },
    {
        id: '2',
        mapHeaderProps: () => {
            return {
                children: 'Heading 2',
            }
        },
        props: {
            prop1: 'prop1-2',
        },
    },
    {
        id: '3',
        props: {
            prop1: 'prop1-3',
        },
    },
    {
        id: '4',
        props: {
            prop1: 'prop1-4',
        },
    },
]

const load = jest.fn()

describe('RichTableContainer', () => {
    beforeEach(() => {
        jest.clearAllMocks()

        useSelectRowLogic.mockImplementation(({
            columns: cp,
        }) => {
            return cp
        })
        useVisible.mockReturnValue([
            columns.map(({
                id,
            }) => {
                return id
            }).filter((_, index) => {
                return index !== 2
            }),
            jest.fn(),
        ])
        useSequence.mockImplementation((paramColumns) => {
            return [
                paramColumns.map(({
                    id,
                }) => { return id }),
                jest.fn(),
            ]
        })
        useFilter.mockReturnValue([
            {
                1: 'val',
            },
            jest.fn(),
            jest.fn(),
        ])
        useSort.mockReturnValue([
            {
                field: '1',
                direction: 'asc',
            },
            jest.fn(),
            jest.fn(),
        ])

        richTableContainerElement = shallow(<RichTableContainer
            load={load}
            name="some-name"
            columns={columns}
            filter={{
                1: {
                    some: 'val',
                },
            }}
        />)
    })

    test('snapshot', () => {
        expect(
            richTableContainerElement,
        ).toMatchSnapshot()
        expect(
            useSequence.mock.calls,
        ).toEqual(
            [[[
                {
                    filterField: 'someField',
                    filterType: 'specialFilter',
                    headerProps: {
                        children: 'Heading 1',
                    },
                    id: '1',
                    label: 'Heading 1',
                    props: {
                        prop1: 'prop1-1',
                    },
                },
                {
                    id: '2',
                    label: 'Heading 2',
                    mapHeaderProps: columns[1].mapHeaderProps,
                    props: {
                        prop1: 'prop1-2',
                    },
                },
                {
                    id: '3',
                    label: '',
                    props: {
                        prop1: 'prop1-3',
                    },
                },
                {
                    id: '4',
                    label: '',
                    props: {
                        prop1: 'prop1-4',
                    },
                },
            ]]],
        )
        expect(
            useVisible.mock.calls,
        ).toEqual(
            [[
                [
                    {
                        filterField: 'someField',
                        filterType: 'specialFilter',
                        headerProps: {
                            children: 'Heading 1',
                        },
                        id: '1',
                        label: 'Heading 1',
                        props: {
                            prop1: 'prop1-1',
                        },
                    },
                    {
                        id: '2',
                        label: 'Heading 2',
                        mapHeaderProps: columns[1].mapHeaderProps,
                        props: {
                            prop1: 'prop1-2',
                        },
                    },
                    {
                        id: '3',
                        label: '',
                        props: {
                            prop1: 'prop1-3',
                        },
                    },
                    {
                        id: '4',
                        label: '',
                        props: {
                            prop1: 'prop1-4',
                        },
                    },
                ],
                [],
            ]],
        )
        expect(
            useFilter.mock.calls,
        ).toEqual(
            [[{
                1: {
                    some: 'val',
                },
            }]],
        )
        expect(
            useSort.mock.calls,
        ).toEqual(
            [[{}]],
        )

        const [[useSelectRowParams]] = useSelectRowLogic.mock.calls

        expect(useSelectRowParams.uniqField).toBe('id')
        expect(useSelectRowParams.onSelectRow).toBe(undefined)
        expect(useSelectRowParams.selectedRows).toEqual(undefined)
        expect(useSelectRowParams.columns.map(({
            id,
        }) => {
            return id
        })).toEqual([
            '1',
            '2',
            '4',
        ])
    })

    test('rerender infinite list', () => {
        let richtableElement = richTableContainerElement.find(RichTable)

        const {
            infinitListKey,
            rerenderInfinitList,
        } = richtableElement.props()

        uuid.mockReturnValue('updated-key')

        rerenderInfinitList()
        richtableElement = richTableContainerElement.find(RichTable)

        const {
            infinitListKey: newInfinitListKey,
        } = richtableElement.props()

        expect(newInfinitListKey).toBe('updated-key')
        expect(infinitListKey).not.toBe(newInfinitListKey)
    })

    test('load with params successfully', () => {
        const {
            load: loadWithParams,
        } = richTableContainerElement.find(RichTable).props()

        load.mockResolvedValue({})

        loadWithParams(3)

        expect(load).toBeCalledWith({
            filter: {
                1: 'val',
            },
            page: 3,
            sort: {
                field: '1',
                direction: 'asc',
            },
            typeMapping: {
                someField: 'specialFilter',
            },
        })
    })

    test('load with params abortive', () => {
        const {
            load: loadWithParams,
        } = richTableContainerElement.find(RichTable).props()

        jest.spyOn(console, 'error')
            .mockImplementation(() => {
                return null
            })

        load.mockRejectedValue({})

        loadWithParams(3)

        expect(load).toBeCalledWith({
            filter: {
                1: 'val',
            },
            page: 3,
            sort: {
                field: '1',
                direction: 'asc',
            },
            typeMapping: {
                someField: 'specialFilter',
            },
        })
    })

    test('setMatchedResults', () => {
        const {
            setMatchedResults,
            matchedResults,
        } = richTableContainerElement.find(RichTable).props()

        expect(matchedResults).toBe(0)

        setMatchedResults(24)

        const {
            matchedResults: updatedMatchedResults,
        } = richTableContainerElement.find(RichTable).props()

        expect(updatedMatchedResults).toBe(24)
    })

    test('sort', () => {
        const [
            _,
            setSortMock,
        ] = useSort()

        const {
            columns: appliedColumns,
        } = richTableContainerElement.find(RichTable).props()

        expect(appliedColumns[1].props.sortDirection).toBe(undefined)

        appliedColumns[1].props.sort('asc')

        expect(setSortMock).toHaveBeenCalledWith('2', 'asc')
    })

    test('filter', () => {
        const [
            _,
            setFilter,
        ] = useFilter()

        const {
            columns: appliedColumns,
        } = richTableContainerElement.find(RichTable).props()

        expect(appliedColumns[1].props.filter).toBe(undefined)

        appliedColumns[1].props.setFilter({
            field: 1,
        })

        expect(setFilter).toHaveBeenCalledWith('2', {
            field: 1,
        })
    })
})
