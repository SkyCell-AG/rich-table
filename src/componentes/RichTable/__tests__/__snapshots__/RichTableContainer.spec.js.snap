// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichTableContainer snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "setDataTable": [Function],
    }
  }
>
  <ForwardRef
    allColumns={
      Array [
        Object {
          "filterField": "someField",
          "filterType": "specialFilter",
          "headerProps": Object {
            "children": "Heading 1",
          },
          "id": "1",
          "label": "Heading 1",
          "props": Object {
            "prop1": "prop1-1",
          },
        },
        Object {
          "id": "2",
          "label": "Heading 2",
          "mapHeaderProps": [Function],
          "props": Object {
            "prop1": "prop1-2",
          },
        },
        Object {
          "id": "3",
          "label": "",
          "props": Object {
            "prop1": "prop1-3",
          },
        },
        Object {
          "id": "4",
          "label": "",
          "props": Object {
            "prop1": "prop1-4",
          },
        },
      ]
    }
    changeSequence={[MockFunction]}
    className=""
    classNames={Object {}}
    columns={
      Array [
        Object {
          "filterField": "someField",
          "filterType": "specialFilter",
          "headerProps": Object {
            "children": "Heading 1",
          },
          "id": "1",
          "label": "Heading 1",
          "props": Object {
            "appliedFilters": Object {
              "includeFilters": Object {
                "1": "val",
              },
            },
            "filter": undefined,
            "predefinedFilter": Object {
              "1": Object {
                "some": "val",
              },
            },
            "setFilter": [Function],
            "sort": [Function],
            "sortDirection": "asc",
          },
        },
        Object {
          "id": "2",
          "label": "Heading 2",
          "mapHeaderProps": [Function],
          "props": Object {
            "appliedFilters": Object {
              "includeFilters": Object {
                "1": "val",
              },
            },
            "filter": undefined,
            "predefinedFilter": Object {
              "1": Object {
                "some": "val",
              },
            },
            "setFilter": [Function],
            "sort": [Function],
            "sortDirection": undefined,
          },
        },
        Object {
          "id": "4",
          "label": "",
          "props": Object {
            "appliedFilters": Object {
              "includeFilters": Object {
                "1": "val",
              },
            },
            "filter": undefined,
            "predefinedFilter": Object {
              "1": Object {
                "some": "val",
              },
            },
            "setFilter": [Function],
            "sort": [Function],
            "sortDirection": undefined,
          },
        },
      ]
    }
    controlPanel={null}
    disabled={false}
    editing={false}
    infinitListKey="some-random-key"
    load={[Function]}
    name="some-name"
    removeFilter={[MockFunction]}
    removeSort={[MockFunction]}
    renderControlPanel={null}
    rerenderInfinitList={[Function]}
    selectedRowId=""
    setVisible={[MockFunction]}
    uniqField="id"
    visible={
      Array [
        "1",
        "2",
        "4",
      ]
    }
  />
</ContextProvider>
`;
